python/cuml/feature_extraction/text.py

```diff
@@ -0,0 +1,473 @@
+# Copyright (c) 2020, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+import warnings
+
+from cudf import Series
+from sklearn.feature_extraction.text import ENGLISH_STOP_WORDS
+from string import punctuation
+from functools import partial
+import nvtext
+import cupy as cp
+import numbers
+
+
+def _preprocess(doc, lower=False, remove_punctuation=False, stop_words=None):
+    """Chain together an optional series of text preprocessing steps to
+    apply to a document.
+    Parameters
+    ----------
+    doc: str
+        The string to preprocess
+    lower: bool
+        Whether to use str.lower to lowercase all of the text
+    remove_punctuation: bool
+        Whether to remove all punctuation from the text before tokenizing.
+        Punctuation characters are taken from string.punctuation
+    Returns
+    -------
+    doc: str
+        preprocessed string
+    """
+    if lower:
+        doc = doc.lower()
+    if remove_punctuation:
+        doc = doc.replace(f'[{punctuation}]', ' ')
+    return doc
+
+
+class _VectorizerMixin:
+    """Provides common code for text vectorizers (tokenization logic)."""
+
+    def _word_ngrams(self, tokens):
+        """Turn tokens into a sequence of n-grams"""
+        min_n, max_n = self.ngram_range
+
+        ngrams = nvtext.ngrams_tokenize(tokens, N=min_n, sep=' ')
+        min_n += 1
+
+        for n in range(min_n, min(max_n + 1, len(tokens) + 1)):
+            ngrams = ngrams.add_strings(
```

Comment:
I wonder if we should have only one call for `add_strings ` rather than n calls to be more efficient.

Reply:
Good point! Working on it.